From 1ef5608571aaa4301554420982d3cce91f965ff6 Mon Sep 17 00:00:00 2001
From: Mika Laitio <lamikr@gmail.com>
Date: Sat, 25 Jan 2025 16:01:30 -0800
Subject: [PATCH 12/13] bitsandbytes TypeError: must be called with a dataclass
 type or instance

bitsandbyte test in

docs/examples/pytorch/bitsandbytes/test_8bit_quantize.py

will cause a following error:

Traceback (most recent call last):
  File "/opt/rocm_sdk_612/lib/python3.11/site-packages/transformers/utils/import_utils.py", line 1817, in _get_module
    return importlib.import_module("." + module_name, self.__name__)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/rocm_sdk_612/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/opt/rocm_sdk_612/lib/python3.11/site-packages/transformers/integrations/bitsandbytes.py", line 21, in <module>
    import bitsandbytes as bnb
  File "/opt/rocm_sdk_612/lib/python3.11/site-packages/bitsandbytes/__init__.py", line 18, in <module>
    from .backends.cpu import CPUBackend
  File "/opt/rocm_sdk_612/lib/python3.11/site-packages/bitsandbytes/backends/cpu.py", line 8, in <module>
    from .cpu_xpu_common import (
  File "/opt/rocm_sdk_612/lib/python3.11/site-packages/bitsandbytes/backends/cpu_xpu_common.py", line 67, in <module>
    @_maybe_torch_compile
     ^^^^^^^^^^^^^^^^^^^^
  File "/opt/rocm_sdk_612/lib/python3.11/site-packages/bitsandbytes/backends/cpu_xpu_common.py", line 63, in _maybe_torch_compile
    return torch.compile(func, dynamic=True, options=options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/rocm_sdk_612/lib/python3.11/site-packages/torch/__init__.py", line 2111, in compile
    backend = _TorchCompileInductorWrapper(mode, options, dynamic)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/rocm_sdk_612/lib/python3.11/site-packages/torch/__init__.py", line 1901, in __init__
    from torch._inductor.compile_fx import compile_fx
  File "/opt/rocm_sdk_612/lib/python3.11/site-packages/torch/_inductor/compile_fx.py", line 42, in <module>
    from torch._inductor.debug import save_args_for_compile_fx_inner
  File "/opt/rocm_sdk_612/lib/python3.11/site-packages/torch/_inductor/debug.py", line 29, in <module>
    from . import config, ir  # noqa: F811, this is needed
    ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/rocm_sdk_612/lib/python3.11/site-packages/torch/_inductor/ir.py", line 72, in <module>
    from .runtime.hints import ReductionHint
  File "/opt/rocm_sdk_612/lib/python3.11/site-packages/torch/_inductor/runtime/hints.py", line 36, in <module>
    attr_desc_fields = {f.name for f in fields(AttrsDescriptor)}
                                        ^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/rocm_sdk_612/lib/python3.11/dataclasses.py", line 1246, in fields
    raise TypeError('must be called with a dataclass type or instance') from None
TypeError: must be called with a dataclass type or instance

Signed-off-by: Mika Laitio <lamikr@gmail.com>
---
 torch/_inductor/runtime/hints.py | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/torch/_inductor/runtime/hints.py b/torch/_inductor/runtime/hints.py
index ba36f40a226..a64a519e2ee 100644
--- a/torch/_inductor/runtime/hints.py
+++ b/torch/_inductor/runtime/hints.py
@@ -38,6 +38,10 @@ try:
     divisible_by_8_available = "divisible_by_8" in attr_desc_fields
 except ImportError:
     attrs_descriptor_available = False
+except TypeError:
+    attrs_descriptor_available = False
+except RuntimeError:
+    attrs_descriptor_available = False
 
 # Define `instance_descriptor` function with clear conditional handling
 if attrs_descriptor_available:
-- 
2.39.5

